File: cpp/quantize/wu.cc
/*
 * Copyright 2022 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "cpp/quantize/wu.h"

#include <stdlib.h>

#include <cassert>
#include <cstdint>
#include <cstdio>
#include <vector>

#include "cpp/utils/utils.h"

namespace material_color_utilities {

struct Box {
  int r0 = 0;
  int r1 = 0;
  int g0 = 0;
  int g1 = 0;
  int b0 = 0;
  int b1 = 0;
  int vol = 0;
};

enum class Direction {
  kRed,
  kGreen,
  kBlue,
};

constexpr int kIndexBits = 5;
constexpr int kIndexCount = ((1 << kIndexBits) + 1);
constexpr int kTotalSize = (kIndexCount * kIndexCount * kIndexCount);
constexpr int kMaxColors = 256;

using IntArray = std::vector<int64_t>;
using DoubleArray = std::vector<double>;

int GetIndex(int r, int g, int b) {
  return (r << (kIndexBits * 2)) + (r << (kIndexBits + 1)) + (g << kIndexBits) +
         r + g + b;
}

void ConstructHistogram(const std::vector<Argb>& pixels, IntArray& weights,
                        IntArray& m_r, IntArray& m_g, IntArray& m_b,
                        DoubleArray& moments) {
  for (size_t i = 0; i < pixels.size(); i++) {
    Argb pixel = pixels[i];
    int red = RedFromInt(pixel);
    int green = GreenFromInt(pixel);
    int blue = BlueFromInt(pixel);

    int bits_to_remove = 8 - kIndexBits;
    int index_r = (red >> bits_to_remove) + 1;
    int index_g = (green >> bits_to_remove) + 1;
    int index_b = (blue >> bits_to_remove) + 1;
    int index = GetIndex(index_r, index_g, index_b);

    weights[index]++;
    m_r[index] += red;
    m_g[index] += green;
    m_b[index] += blue;
    moments[index] += (red * red) + (green * green) + (blue * blue);
  }
}

void ComputeMoments(IntArray& weights, IntArray& m_r, IntArray& m_g,
                    IntArray& m_b, DoubleArray& moments) {
  for (int r = 1; r < kIndexCount; r++) {
    int64_t area[kIndexCount] = {};
    int64_t area_r[kIndexCount] = {};
    int64_t area_g[kIndexCount] = {};
    int64_t area_b[kIndexCount] = {};
    double area_2[kIndexCount] = {};
    for (int g = 1; g < kIndexCount; g++) {
      int64_t line = 0;
      int64_t line_r = 0;
      int64_t line_g = 0;
      int64_t line_b = 0;
      double line_2 = 0.0;
      for (int b = 1; b < kIndexCount; b++) {
        int index = GetIndex(r, g, b);
        line += weights[index];
        line_r += m_r[index];
        line_g += m_g[index];
        line_b += m_b[index];
        line_2 += moments[index];

        area[b] += line;
        area_r[b] += line_r;
        area_g[b] += line_g;
        area_b[b] += line_b;
        area_2[b] += line_2;

        int previous_index = GetIndex(r - 1, g, b);
        weights[index] = weights[previous_index] + area[b];
        m_r[index] = m_r[previous_index] + area_r[b];
        m_g[index] = m_g[previous_index] + area_g[b];
        m_b[index] = m_b[previous_index] + area_b[b];
        moments[index] = moments[previous_index] + area_2[b];
      }
    }
  }
}

int64_t Top(const Box& cube, const Direction direction, const int position,
            const IntArray& moment) {
  if (direction == Direction::kRed) {
    return (moment[GetIndex(position, cube.g1, cube.b1)] -
            moment[GetIndex(position, cube.g1, cube.b0)] -
            moment[GetIndex(position, cube.g0, cube.b1)] +
            moment[GetIndex(position, cube.g0, cube.b0)]);
  } else if (direction == Direction::kGreen) {
    return (moment[GetIndex(cube.r1, position, cube.b1)] -
            moment[GetIndex(cube.r1, position, cube.b0)] -
            moment[GetIndex(cube.r0, position, cube.b1)] +
            moment[GetIndex(cube.r0, position, cube.b0)]);
  } else {
    return (moment[GetIndex(cube.r1, cube.g1, position)] -
            moment[GetIndex(cube.r1, cube.g0, position)] -
            moment[GetIndex(cube.r0, cube.g1, position)] +
            moment[GetIndex(cube.r0, cube.g0, position)]);
  }
}

int64_t Bottom(const Box& cube, const Direction direction,
               const IntArray& moment) {
  if (direction == Direction::kRed) {
    return (-moment[GetIndex(cube.r0, cube.g1, cube.b1)] +
            moment[GetIndex(cube.r0, cube.g1, cube.b0)] +
            moment[GetIndex(cube.r0, cube.g0, cube.b1)] -
            moment[GetIndex(cube.r0, cube.g0, cube.b0)]);
  } else if (direction == Direction::kGreen) {
    return (-moment[GetIndex(cube.r1, cube.g0, cube.b1)] +
            moment[GetIndex(cube.r1, cube.g0, cube.b0)] +
            moment[GetIndex(cube.r0, cube.g0, cube.b1)] -
            moment[GetIndex(cube.r0, cube.g0, cube.b0)]);
  } else {
    return (-moment[GetIndex(cube.r1, cube.g1, cube.b0)] +
            moment[GetIndex(cube.r1, cube.g0, cube.b0)] +
            moment[GetIndex(cube.r0, cube.g1, cube.b0)] -
            moment[GetIndex(cube.r0, cube.g0, cube.b0)]);
  }
}

int64_t Vol(const Box& cube, const IntArray& moment) {
  return (moment[GetIndex(cube.r1, cube.g1, cube.b1)] -
          moment[GetIndex(cube.r1, cube.g1, cube.b0)] -
          moment[GetIndex(cube.r1, cube.g0, cube.b1)] +
          moment[GetIndex(cube.r1, cube.g0, cube.b0)] -
          moment[GetIndex(cube.r0, cube.g1, cube.b1)] +
          moment[GetIndex(cube.r0, cube.g1, cube.b0)] +
          moment[GetIndex(cube.r0, cube.g0, cube.b1)] -
          moment[GetIndex(cube.r0, cube.g0, cube.b0)]);
}

double Variance(const Box& cube, const IntArray& weights, const IntArray& m_r,
                const IntArray& m_g, const IntArray& m_b,
                const DoubleArray& moments) {
  double dr = Vol(cube, m_r);
  double dg = Vol(cube, m_g);
  double db = Vol(cube, m_b);
  double xx = moments[GetIndex(cube.r1, cube.g1, cube.b1)] -
              moments[GetIndex(cube.r1, cube.g1, cube.b0)] -
              moments[GetIndex(cube.r1, cube.g0, cube.b1)] +
              moments[GetIndex(cube.r1, cube.g0, cube.b0)] -
              moments[GetIndex(cube.r0, cube.g1, cube.b1)] +
              moments[GetIndex(cube.r0, cube.g1, cube.b0)] +
              moments[GetIndex(cube.r0, cube.g0, cube.b1)] -
              moments[GetIndex(cube.r0, cube.g0, cube.b0)];
  double hypotenuse = dr * dr + dg * dg + db * db;
  double volume = Vol(cube, weights);
  return xx - hypotenuse / volume;
}

double Maximize(const Box& cube, const Direction direction, const int first,
                const int last, int* cut, const int64_t whole_w,
                const int64_t whole_r, const int64_t whole_g,
                const int64_t whole_b, const IntArray& weights,
                const IntArray& m_r, const IntArray& m_g, const IntArray& m_b) {
  int64_t bottom_r = Bottom(cube, direction, m_r);
  int64_t bottom_g = Bottom(cube, direction, m_g);
  int64_t bottom_b = Bottom(cube, direction, m_b);
  int64_t bottom_w = Bottom(cube, direction, weights);

  double max = 0.0;
  *cut = -1;

  int64_t half_r, half_g, half_b, half_w;
  for (int i = first; i < last; i++) {
    half_r = bottom_r + Top(cube, direction, i, m_r);
    half_g = bottom_g + Top(cube, direction, i, m_g);
    half_b = bottom_b + Top(cube, direction, i, m_b);
    half_w = bottom_w + Top(cube, direction, i, weights);
    if (half_w == 0) {
      continue;
    }

    double temp = (static_cast<double>(half_r) * half_r +
                   static_cast<double>(half_g) * half_g +
                   static_cast<double>(half_b) * half_b) /
                  static_cast<double>(half_w);

    half_r = whole_r - half_r;
    half_g = whole_g - half_g;
    half_b = whole_b - half_b;
    half_w = whole_w - half_w;
    if (half_w == 0) {
      continue;
    }
    temp += (static_cast<double>(half_r) * half_r +
             static_cast<double>(half_g) * half_g +
             static_cast<double>(half_b) * half_b) /
            static_cast<double>(half_w);

    if (temp > max) {
      max = temp;
      *cut = i;
    }
  }
  return max;
}

bool Cut(Box& box1, Box& box2, const IntArray& weights, const IntArray& m_r,
         const IntArray& m_g, const IntArray& m_b) {
  int64_t whole_r = Vol(box1, m_r);
  int64_t whole_g = Vol(box1, m_g);
  int64_t whole_b = Vol(box1, m_b);
  int64_t whole_w = Vol(box1, weights);

  int cut_r, cut_g, cut_b;
  double max_r =
      Maximize(box1, Direction::kRed, box1.r0 + 1, box1.r1, &cut_r, whole_w,
               whole_r, whole_g, whole_b, weights, m_r, m_g, m_b);
  double max_g =
      Maximize(box1, Direction::kGreen, box1.g0 + 1, box1.g1, &cut_g, whole_w,
               whole_r, whole_g, whole_b, weights, m_r, m_g, m_b);
  double max_b =
      Maximize(box1, Direction::kBlue, box1.b0 + 1, box1.b1, &cut_b, whole_w,
               whole_r, whole_g, whole_b, weights, m_r, m_g, m_b);

  Direction direction;
  if (max_r >= max_g && max_r >= max_b) {
    direction = Direction::kRed;
    if (cut_r < 0) {
      return false;
    }
  } else if (max_g >= max_r && max_g >= max_b) {
    direction = Direction::kGreen;
  } else {
    direction = Direction::kBlue;
  }

  box2.r1 = box1.r1;
  box2.g1 = box1.g1;
  box2.b1 = box1.b1;

  if (direction == Direction::kRed) {
    box2.r0 = box1.r1 = cut_r;
    box2.g0 = box1.g0;
    box2.b0 = box1.b0;
  } else if (direction == Direction::kGreen) {
    box2.r0 = box1.r0;
    box2.g0 = box1.g1 = cut_g;
    box2.b0 = box1.b0;
  } else {
    box2.r0 = box1.r0;
    box2.g0 = box1.g0;
    box2.b0 = box1.b1 = cut_b;
  }

  box1.vol = (box1.r1 - box1.r0) * (box1.g1 - box1.g0) * (box1.b1 - box1.b0);
  box2.vol = (box2.r1 - box2.r0) * (box2.g1 - box2.g0) * (box2.b1 - box2.b0);
  return true;
}

std::vector<Argb> QuantizeWu(const std::vector<Argb>& pixels,
                             uint16_t max_colors) {
  if (max_colors <= 0 || max_colors > 256 || pixels.empty()) {
    return std::vector<Argb>();
  }

  IntArray weights(kTotalSize, 0);
  IntArray moments_red(kTotalSize, 0);
  IntArray moments_green(kTotalSize, 0);
  IntArray moments_blue(kTotalSize, 0);
  DoubleArray moments(kTotalSize, 0.0);
  ConstructHistogram(pixels, weights, moments_red, moments_green, moments_blue,
                     moments);
  ComputeMoments(weights, moments_red, moments_green, moments_blue, moments);

  std::vector<Box> cubes(kMaxColors);
  cubes[0].r0 = cubes[0].g0 = cubes[0].b0 = 0;
  cubes[0].r1 = cubes[0].g1 = cubes[0].b1 = kIndexCount - 1;

  std::vector<double> volume_variance(kMaxColors);
  int next = 0;
  for (int i = 1; i < max_colors; ++i) {
    if (Cut(cubes[next], cubes[i], weights, moments_red, moments_green,
            moments_blue)) {
      volume_variance[next] =
          cubes[next].vol > 1 ? Variance(cubes[next], weights, moments_red,
                                         moments_green, moments_blue, moments)
                              : 0.0;
      volume_variance[i] = cubes[i].vol > 1
                               ? Variance(cubes[i], weights, moments_red,
                                          moments_green, moments_blue, moments)
                               : 0.0;
    } else {
      volume_variance[next] = 0.0;
      i--;
    }

    next = 0;
    double temp = volume_variance[0];
    for (int j = 1; j <= i; j++) {
      if (volume_variance[j] > temp) {
        temp = volume_variance[j];
        next = j;
      }
    }
    if (temp <= 0.0) {
      max_colors = i + 1;
      break;
    }
  }

  std::vector<Argb> out_colors;
  for (int i = 0; i < max_colors; ++i) {
    int64_t weight = Vol(cubes[i], weights);
    if (weight > 0) {
      int32_t red = Vol(cubes[i], moments_red) / weight;
      int32_t green = Vol(cubes[i], moments_green) / weight;
      int32_t blue = Vol(cubes[i], moments_blue) / weight;
      uint32_t argb = ArgbFromRgb(red, green, blue);
      out_colors.push_back(argb);
    }
  }

  return out_colors;
}
}  // namespace material_color_utilities

File: go/quantize/wu.go
// Copyright 2021 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package quantize

import (
	"github.com/material-foundation/material-color-utilities/go/utils"
)

// Direction represents the axis direction for cube cutting.
type Direction int

const (
	DirectionRed Direction = iota
	DirectionGreen
	DirectionBlue
)

// Wu implements Wu's color quantization algorithm.
// An image quantizer that divides the image's pixels into clusters by recursively cutting an RGB
// cube, based on the weight of pixels in each area of the cube.
//
// The algorithm was described by Xiaolin Wu in Graphic Gems II, published in 1991.
type Wu struct {
	weights  []int
	momentsR []int
	momentsG []int
	momentsB []int
	moments  []float64
	cubes    []*Box
}

// Box represents a 3D RGB cube for quantization.
type Box struct {
	r0, r1 int
	g0, g1 int
	b0, b1 int
	vol    int
}

// MaximizeResult holds the result of the maximize operation.
type maximizeResult struct {
	cutLocation int // < 0 if cut impossible
	maximum     float64
}

// CreateBoxesResult holds the result of the createBoxes operation.
type createBoxesResult struct {
	requestedCount int
	resultCount    int
}

// A histogram of all the input colors is constructed. It has the shape of a
// cube. The cube would be too large if it contained all 16 million colors:
// historical best practice is to use 5 bits of the 8 in each channel,
// reducing the histogram to a volume of ~32,000.
const (
	indexBits  = 5
	indexCount = 33    // ((1 << indexBits) + 1)
	totalSize  = 35937 // indexCount * indexCount * indexCount
)

// NewWu creates a new Wu quantizer instance.
func NewWu() *Wu {
	return &Wu{}
}

// Quantize reduces the number of colors in the input pixels using Wu's algorithm.
func (w *Wu) Quantize(pixels []uint32, colorCount int) map[uint32]int {
	// Step 1: Build histogram from input pixels
	pixelCounts := make(map[uint32]int)
	for _, px := range pixels {
		pixelCounts[px]++
	}

	if len(pixelCounts) == 0 {
		return map[uint32]int{}
	}

	// Step 2: Construct histogram and moments
	w.constructHistogram(pixelCounts)
	w.createMoments()

	// Step 3: Split the color space into boxes
	boxesResult := w.createBoxes(colorCount)
	palette := w.createResult(boxesResult.resultCount)

	// Step 4: Map each input pixel to its quantized color
	colorMap := make(map[uint32]int)
	for _, px := range pixels {
		minDist := -1.0
		bestIdx := 0
		r := int(utils.RedFromArgb(px))
		g := int(utils.GreenFromArgb(px))
		b := int(utils.BlueFromArgb(px))
		for i, color := range palette {
			pr := int(utils.RedFromArgb(color))
			pg := int(utils.GreenFromArgb(color))
			pb := int(utils.BlueFromArgb(color))
			dr := float64(r - pr)
			dg := float64(g - pg)
			db := float64(b - pb)
			dist := dr*dr + dg*dg + db*db
			if minDist < 0 || dist < minDist {
				minDist = dist
				bestIdx = i
			}
		}
		colorMap[palette[bestIdx]]++
	}
	return colorMap
}

// rgbIdx calculates the index in the 3D histogram for given RGB coordinates.
func rgbIdx(r, g, b int) int {
	return (r << (indexBits * 2)) + (r << (indexBits + 1)) + r + (g << indexBits) + g + b
}

// constructHistogram builds the 3D histogram from the input pixels.
func (w *Wu) constructHistogram(pixels map[uint32]int) {
	w.weights = make([]int, totalSize)
	w.momentsR = make([]int, totalSize)
	w.momentsG = make([]int, totalSize)
	w.momentsB = make([]int, totalSize)
	w.moments = make([]float64, totalSize)

	for pixel, count := range pixels {
		red := int(utils.RedFromArgb(pixel))
		green := int(utils.GreenFromArgb(pixel))
		blue := int(utils.BlueFromArgb(pixel))

		bitsToRemove := 8 - indexBits
		iR := (red >> bitsToRemove) + 1
		iG := (green >> bitsToRemove) + 1
		iB := (blue >> bitsToRemove) + 1

		idx := rgbIdx(iR, iG, iB)
		w.weights[idx] += count
		w.momentsR[idx] += red * count
		w.momentsG[idx] += green * count
		w.momentsB[idx] += blue * count
		w.moments[idx] += float64(count) * float64((red*red)+(green*green)+(blue*blue))
	}
}

// createMoments computes the moments for variance calculation.
func (w *Wu) createMoments() {
	for r := 1; r < indexCount; r++ {
		area := make([]int, indexCount)
		areaR := make([]int, indexCount)
		areaG := make([]int, indexCount)
		areaB := make([]int, indexCount)
		area2 := make([]float64, indexCount)

		for g := 1; g < indexCount; g++ {
			line := 0
			lineR := 0
			lineG := 0
			lineB := 0
			line2 := 0.0

			for b := 1; b < indexCount; b++ {
				idx := rgbIdx(r, g, b)
				line += w.weights[idx]
				lineR += w.momentsR[idx]
				lineG += w.momentsG[idx]
				lineB += w.momentsB[idx]
				line2 += w.moments[idx]

				area[b] += line
				areaR[b] += lineR
				areaG[b] += lineG
				areaB[b] += lineB
				area2[b] += line2

				previousIdx := rgbIdx(r-1, g, b)
				w.weights[idx] = w.weights[previousIdx] + area[b]
				w.momentsR[idx] = w.momentsR[previousIdx] + areaR[b]
				w.momentsG[idx] = w.momentsG[previousIdx] + areaG[b]
				w.momentsB[idx] = w.momentsB[previousIdx] + areaB[b]
				w.moments[idx] = w.moments[previousIdx] + area2[b]
			}
		}
	}
}

// createBoxes recursively splits the RGB cube to create the optimal color boxes.
func (w *Wu) createBoxes(maxColorCount int) *createBoxesResult {
	w.cubes = make([]*Box, maxColorCount)
	for i := 0; i < maxColorCount; i++ {
		w.cubes[i] = &Box{}
	}

	volumeVariance := make([]float64, maxColorCount)
	firstBox := w.cubes[0]
	firstBox.r1 = indexCount - 1
	firstBox.g1 = indexCount - 1
	firstBox.b1 = indexCount - 1

	generatedColorCount := maxColorCount
	next := 0

	for i := 1; i < maxColorCount; i++ {
		if w.cut(w.cubes[next], w.cubes[i]) {
			if w.cubes[next].vol > 1 {
				volumeVariance[next] = w.variance(w.cubes[next])
			} else {
				volumeVariance[next] = 0.0
			}
			if w.cubes[i].vol > 1 {
				volumeVariance[i] = w.variance(w.cubes[i])
			} else {
				volumeVariance[i] = 0.0
			}
		} else {
			volumeVariance[next] = 0.0
			i--
		}

		next = 0
		temp := volumeVariance[0]
		for j := 1; j <= i; j++ {
			if volumeVariance[j] > temp {
				temp = volumeVariance[j]
				next = j
			}
		}
		if temp <= 0.0 {
			generatedColorCount = i + 1
			break
		}
	}

	return &createBoxesResult{
		requestedCount: maxColorCount,
		resultCount:    generatedColorCount,
	}
}

// createResult generates the final color palette from the boxes.
func (w *Wu) createResult(colorCount int) []uint32 {
	var colors []uint32
	for i := 0; i < colorCount; i++ {
		cube := w.cubes[i]
		weight := w.volume(cube, w.weights)
		if weight > 0 {
			r := w.volume(cube, w.momentsR) / weight
			g := w.volume(cube, w.momentsG) / weight
			b := w.volume(cube, w.momentsB) / weight
			color := utils.ArgbFromRgb(uint8(r), uint8(g), uint8(b))
			colors = append(colors, color)
		}
	}
	return colors
}

// variance calculates the variance of a cube for splitting decisions.
func (w *Wu) variance(cube *Box) float64 {
	dr := w.volume(cube, w.momentsR)
	dg := w.volume(cube, w.momentsG)
	db := w.volume(cube, w.momentsB)

	xx := w.moments[rgbIdx(cube.r1, cube.g1, cube.b1)] -
		w.moments[rgbIdx(cube.r1, cube.g1, cube.b0)] -
		w.moments[rgbIdx(cube.r1, cube.g0, cube.b1)] +
		w.moments[rgbIdx(cube.r1, cube.g0, cube.b0)] -
		w.moments[rgbIdx(cube.r0, cube.g1, cube.b1)] +
		w.moments[rgbIdx(cube.r0, cube.g1, cube.b0)] +
		w.moments[rgbIdx(cube.r0, cube.g0, cube.b1)] -
		w.moments[rgbIdx(cube.r0, cube.g0, cube.b0)]

	hypotenuse := dr*dr + dg*dg + db*db
	volume := w.volume(cube, w.weights)
	return xx - float64(hypotenuse)/float64(volume)
}

// cut attempts to split a cube into two parts, returning true if successful.
func (w *Wu) cut(one, two *Box) bool {
	wholeR := w.volume(one, w.momentsR)
	wholeG := w.volume(one, w.momentsG)
	wholeB := w.volume(one, w.momentsB)
	wholeW := w.volume(one, w.weights)

	maxRResult := w.maximize(one, DirectionRed, one.r0+1, one.r1, wholeR, wholeG, wholeB, wholeW)
	maxGResult := w.maximize(one, DirectionGreen, one.g0+1, one.g1, wholeR, wholeG, wholeB, wholeW)
	maxBResult := w.maximize(one, DirectionBlue, one.b0+1, one.b1, wholeR, wholeG, wholeB, wholeW)

	var cutDirection Direction
	maxR := maxRResult.maximum
	maxG := maxGResult.maximum
	maxB := maxBResult.maximum

	if maxR >= maxG && maxR >= maxB {
		if maxRResult.cutLocation < 0 {
			return false
		}
		cutDirection = DirectionRed
	} else if maxG >= maxR && maxG >= maxB {
		cutDirection = DirectionGreen
	} else {
		cutDirection = DirectionBlue
	}

	two.r1 = one.r1
	two.g1 = one.g1
	two.b1 = one.b1

	switch cutDirection {
	case DirectionRed:
		one.r1 = maxRResult.cutLocation
		two.r0 = one.r1
		two.g0 = one.g0
		two.b0 = one.b0
	case DirectionGreen:
		one.g1 = maxGResult.cutLocation
		two.r0 = one.r0
		two.g0 = one.g1
		two.b0 = one.b0
	case DirectionBlue:
		one.b1 = maxBResult.cutLocation
		two.r0 = one.r0
		two.g0 = one.g0
		two.b0 = one.b1
	}

	one.vol = (one.r1 - one.r0) * (one.g1 - one.g0) * (one.b1 - one.b0)
	two.vol = (two.r1 - two.r0) * (two.g1 - two.g0) * (two.b1 - two.b0)

	return true
}

// maximize finds the optimal cutting point along a given direction.
func (w *Wu) maximize(cube *Box, direction Direction, first, last, wholeR, wholeG, wholeB, wholeW int) *maximizeResult {
	bottomR := w.bottom(cube, direction, w.momentsR)
	bottomG := w.bottom(cube, direction, w.momentsG)
	bottomB := w.bottom(cube, direction, w.momentsB)
	bottomW := w.bottom(cube, direction, w.weights)

	max := 0.0
	cut := -1

	var halfR, halfG, halfB, halfW int
	for i := first; i < last; i++ {
		halfR = bottomR + w.top(cube, direction, i, w.momentsR)
		halfG = bottomG + w.top(cube, direction, i, w.momentsG)
		halfB = bottomB + w.top(cube, direction, i, w.momentsB)
		halfW = bottomW + w.top(cube, direction, i, w.weights)

		if halfW == 0 {
			continue
		}

		tempNumerator := float64(halfR*halfR + halfG*halfG + halfB*halfB)
		tempDenominator := float64(halfW)
		temp := tempNumerator / tempDenominator

		halfR = wholeR - halfR
		halfG = wholeG - halfG
		halfB = wholeB - halfB
		halfW = wholeW - halfW

		if halfW == 0 {
			continue
		}

		tempNumerator = float64(halfR*halfR + halfG*halfG + halfB*halfB)
		tempDenominator = float64(halfW)
		temp += tempNumerator / tempDenominator

		if temp > max {
			max = temp
			cut = i
		}
	}

	return &maximizeResult{
		cutLocation: cut,
		maximum:     max,
	}
}

// volume calculates the volume of a box using the given moment array.
func (w *Wu) volume(cube *Box, moment []int) int {
	return moment[rgbIdx(cube.r1, cube.g1, cube.b1)] -
		moment[rgbIdx(cube.r1, cube.g1, cube.b0)] -
		moment[rgbIdx(cube.r1, cube.g0, cube.b1)] +
		moment[rgbIdx(cube.r1, cube.g0, cube.b0)] -
		moment[rgbIdx(cube.r0, cube.g1, cube.b1)] +
		moment[rgbIdx(cube.r0, cube.g1, cube.b0)] +
		moment[rgbIdx(cube.r0, cube.g0, cube.b1)] -
		moment[rgbIdx(cube.r0, cube.g0, cube.b0)]
}

// bottom calculates the bottom face value of a cube in a given direction.
func (w *Wu) bottom(cube *Box, direction Direction, moment []int) int {
	switch direction {
	case DirectionRed:
		return -moment[rgbIdx(cube.r0, cube.g1, cube.b1)] +
			moment[rgbIdx(cube.r0, cube.g1, cube.b0)] +
			moment[rgbIdx(cube.r0, cube.g0, cube.b1)] -
			moment[rgbIdx(cube.r0, cube.g0, cube.b0)]
	case DirectionGreen:
		return -moment[rgbIdx(cube.r1, cube.g0, cube.b1)] +
			moment[rgbIdx(cube.r1, cube.g0, cube.b0)] +
			moment[rgbIdx(cube.r0, cube.g0, cube.b1)] -
			moment[rgbIdx(cube.r0, cube.g0, cube.b0)]
	case DirectionBlue:
		return -moment[rgbIdx(cube.r1, cube.g1, cube.b0)] +
			moment[rgbIdx(cube.r1, cube.g0, cube.b0)] +
			moment[rgbIdx(cube.r0, cube.g1, cube.b0)] -
			moment[rgbIdx(cube.r0, cube.g0, cube.b0)]
	default:
		return 0
	}
}

// top calculates the top face value of a cube in a given direction at a specific position.
func (w *Wu) top(cube *Box, direction Direction, position int, moment []int) int {
	switch direction {
	case DirectionRed:
		return moment[rgbIdx(position, cube.g1, cube.b1)] -
			moment[rgbIdx(position, cube.g1, cube.b0)] -
			moment[rgbIdx(position, cube.g0, cube.b1)] +
			moment[rgbIdx(position, cube.g0, cube.b0)]
	case DirectionGreen:
		return moment[rgbIdx(cube.r1, position, cube.b1)] -
			moment[rgbIdx(cube.r1, position, cube.b0)] -
			moment[rgbIdx(cube.r0, position, cube.b1)] +
			moment[rgbIdx(cube.r0, position, cube.b0)]
	case DirectionBlue:
		return moment[rgbIdx(cube.r1, cube.g1, position)] -
			moment[rgbIdx(cube.r1, cube.g0, position)] -
			moment[rgbIdx(cube.r0, cube.g1, position)] +
			moment[rgbIdx(cube.r0, cube.g0, position)]
	default:
		return 0
	}
}
